new purchase department modules

lepistone opened the pull request:
dependencies:
- [x] OCA/purchase-workflow#70

lepistone committed:
new module purchase_department, with tree view

lepistone committed:
add department to po form view

lepistone committed:
add department in po search view

lepistone committed:
department in purchase order line

pedrobaeza commented:
Thanks for the module, Leonardo.

I think you should leave purchase order line department field as regular, because you can encode purchases together in one order, but corresponding to different departments in the cost center structure of the company. Department in header then will be the default for new lines.

lepistone commented:
Hi Pedro! FYI, the `purchase_department` is done, but I marked the PR as WIP because I'm about to add a similar little module for tenders.

As for your question, I'm not sure. I'll ping @gurneyalex and @jgrandguillaume for they opinion on this.

Thanks!

lepistone committed:
new module purchase_requisition_department

lepistone committed:
tender form view

lepistone committed:
add requisition search view

gurneyalex commented:
@pedrobaeza I concerned about issues happening with a structure such as 

depA
 |- depB
 |-depC

Sally from depB creates a PO and adds lines. 
Bob from depA adds a line to the PO and says it's related to depC
Sally confirms PO and gets errors because she cannot read the line related to depC (if ACL are added)

So I'd rather have the current behavior and let people override it in a custom module if needed.

lepistone committed:
no department on tender lines: there are no views

lepistone committed:
propagate department from tender to bid

lepistone committed:
propagate empty department to invoice

lepistone committed:
actually propagate department id to the invoice

lepistone committed:
lint

lepistone committed:
improve readme

pedrobaeza commented:
@gurneyalex, by default, on core you don't have any record rules on departments, so that's not an issue.

lepistone commented:
@pedrobaeza but those modules are there mainly to allow you to add record rules on objects, so I think that mixed lines can indeed give trouble.

lepistone committed:
travis: clone purchase-workflow

lepistone committed:
framework_agreement_department

This module is just like the others here: adds a notion of department to
agreements.

pedrobaeza commented:
But the limitation must come then in the customized module that adds that record rules, not here.

lepistone commented:
From a technical point of view, I agree with you that we often use this pattern where lines have a default that is taken from the header. 

However, the use case I have in mind, where departments are used to classify documents, makes departments on lines confusing and error-prone.

For example, even without ACLs, group-by's and filters are added to the search views, and I see much confusion if users open an order and see lines belonging to another department.

The only reason for which I added the department to lines is just to show it to the users where the line has its own menuitem and views (like purchase order lines).

But I understand and respect your point Pedro. I'll ping again @gurneyalex and @jgrandguillaume and other reviewers, and if the default way prevails I'll change as you do and override somewhere else.

Thanks!

jgrandguillaume commented:
Hi there,

First, thanks for the review and contribs ! My point of view on this is that we always added department on "main objects" not "lines". For example on invoice, on sale order...

So, having department on purchase.order is ok, reporting that infos on line allow to use "po line menu" sounds good to me. If we put department on po lines, then it'll become inconsistent with invoice and sales... e.g. What department is an invoice issue from a PO linked to 3 department ?

Moreover, I think in most cases, people work and validate a document (like PO) within a department, because budget are linked to them. So it seems a bit strange to have a PO with various departments, as it'll be a bit difficult to know who must validate...

I suggest to stay with the same line that we are know: managing department at "main object" level.

In a second round, we may add "invoice_line_department", "sale_line_department",.... to deal with that if the need raise..

So for me :+1: 

Regards

yvaucher commented on purchase_requisition_department/LICENSE:
```diff
@@ -0,0 +1,17 @@
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public Lice
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
```
About license: http://www.gnu.org/licenses/agpl-3.0.html
I think this won't be enough that way.

In `How to Apply These Terms to Your New Programs`
`each file should have at least the "copyright" line and a pointer to where the full notice is found.`

lepistone replied:
done, thanks

pedrobaeza commented:
OK, thanks for the explanations. If you see department as cost center, I still see needed the split by lines, but I don't insist more. As you have said, this can come via extension modules.

Please fix Travis and you will get my :+1:

lepistone committed:
travis: pull server-tools

lepistone committed:
remove unused import

lepistone committed:
move copyright notice to all python files

lepistone commented:
Thanks @pedrobaeza! We're now green.

yvaucher commented on purchase_requisition_department/model/purchase.py:
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
Missing license

pedrobaeza replied:
What I'm doing lastly is to put:

```
##############################################################################
# For copyright and license notices, see __openerp__.py file in root directory
##############################################################################
```

yvaucher replied:
@pedrobaeza I think it still miss the copyright line from what I can read in http://www.gnu.org/licenses/agpl-3.0.html.

You still needs something like:

```
Copyright (C) <year>  <name of author> <where full licence notice can be found>
```

pedrobaeza replied:
The link talks about suggestions, but not mandatory. I think that if the copyright is the same, is more maintenable to have it in one place, and update that place when someone else add something. Or do you add yourself in each file that you touch as a contribution?

lepistone replied:
I think for that PR I'll do as usual with headers in every file. Otherwise, if it's legally sound I'm happy to reduce those headers to the minimum in the future. 

We could discuss that in the list.

lepistone commented:
I moved the license on all python files for the three modules in this PR.

As said, I'll support a general proposal to reduce headers as much as is legally sound. Thanks!

yvaucher commented:
:+1:
